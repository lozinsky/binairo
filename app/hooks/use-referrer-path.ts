import { parsePath, useLocation } from '@remix-run/react';
import { useMemo } from 'react';

export function useReferrerPath() {
  const location = useLocation();
  const referrerPath = useMemo(() => {
    const referrer = new URLSearchParams(location.search).get('referrer');

    if (referrer === null) {
      return null;
    }

    return parsePath(referrer);
  }, [location.search]);

  return referrerPath;
}
