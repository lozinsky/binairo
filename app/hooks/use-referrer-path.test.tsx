/* @vitest-environment jsdom */

import type { Location } from '@remix-run/react';
import type { ReactNode } from 'react';

import { createRemixStub } from '@remix-run/testing';
import { renderHook } from '@testing-library/react';
import { expect, test } from 'vitest';

import { useReferrerPath } from './use-referrer-path';

test.each<{ location: Partial<Location> }>([
  {
    location: {
      search: '',
    },
  },
  {
    location: {
      search: '?referrer=%2F',
    },
  },
  {
    location: {
      search: '?referrer=%2Fgame%3Freferrer%3D%252F',
    },
  },
  {
    location: {
      search: `?referrer=%2Fgame%2FW1tbMSwyXSxbMSwyXV0sW1swLDBdLFswLDFdXV0%3D%3Freferrer%3D%252Fgame%253Freferrer%253D%25252F`,
    },
  },
])('returns referrer path', ({ location }) => {
  function Wrapper({ children }: { children: ReactNode }) {
    const RemixStub = createRemixStub([{ Component, path: '/' }]);

    function Component() {
      return children;
    }

    return <RemixStub initialEntries={[location]} />;
  }

  const { result } = renderHook(
    () => {
      const referrerPath = useReferrerPath();

      return referrerPath;
    },
    { wrapper: Wrapper },
  );

  expect(result.current).toMatchSnapshot();
});
