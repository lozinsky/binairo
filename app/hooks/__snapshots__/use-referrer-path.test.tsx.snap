// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns referrer path 1`] = `null`;

exports[`returns referrer path 2`] = `
{
  "pathname": "/",
}
`;

exports[`returns referrer path 3`] = `
{
  "pathname": "/game",
  "search": "?referrer=%2F",
}
`;

exports[`returns referrer path 4`] = `
{
  "pathname": "/game/W1tbMSwyXSxbMSwyXV0sW1swLDBdLFswLDFdXV0=",
  "search": "?referrer=%2Fgame%3Freferrer%3D%252F",
}
`;
